### ISSUE TITLE
Use MathJax to typeset TEX svg images in ZIM files (at least in jQuery mode)

### ISSUE THREAD

**Issue by Jaifroid** [kiwix/kiwix-js#297]
Display of svg images is slow and processor intensive, and has the potential for causing the browser to hang (especially Chrome). Many svg images in Wikipedia files are fallback images for TEX-based mathematical formulae. The TEX string is usually provided in the image's alt text attribute. Therefore it would make sense to process this TEX string instead and typeset using the cross-browser, robust maths typesetting JavaScript library [MathJax](https://www.mathjax.org/). This could be an option especially for jQuery mode, but it could also be beneficial for ServiceWorker mode.

**Comment by Jaifroid** [comment]
Commit https://github.com/kiwix/kiwix-js/commit/d5ee02bfbbaf0d89307663853d3d0164ee6b5121 adds experimental support for this. Tested in Edge, Firefox, Edge Mobile, UWP and Firefox OS.

The MathJax library is loaded dynamically if needed. Note that on first use on a heavy page, it can be slow to set itself up in FFOS, but on subsequent scrolls and with subsequent mathematical page, equations are displayed very fast. There may be a remedy in attempting to typeset fewer images on first pass.

The subroutine falls back to using SVGs if there is no TEX string in the alt field, or if the field uses PNG images rather than SVGs, as there is no advantage in that case.

Screenshot below is FFOS using the latest Fench Wikipedia ZIM `wikipedia_fr_all_novid_2017-07.zim`:

![image](https://user-images.githubusercontent.com/4304337/29016668-3b21ed1a-7b4b-11e7-866d-157526de3e09.png)

**Comment by Jaifroid** [comment]
TODO: User regex pattern to detect that the string in the alt field truly is TEX. Backslash detection should be sufficient, though the class name `mwe-math-fallback-image` is usually sufficient guarantee that if there is alt text, it is a TEX string.

**Comment by kelson42** [comment]
@Jaifroid Everybody goes to Mathjax. Wikipedia too. That is, I agree, a good move. That said, this should really be implemented in ZIM file itself. Any chance you would move the effort you are currently providing for kiwix-js to mwoffliner (this is js too)?

BTW, We already have ZIM file with Mathjax, the one of StackExchange.

**Comment by Jaifroid** [comment]
@kelson42 Thank you for the comment. I'd be very interested to see how it is implemented in the ZIM file directly and what that looks like.

Presumably some support is still needed in the client, as MathJax is an interpreter of the TEX string and does the display at client level using CSS and its own mathematical fonts.

Once I've got a kiwix-js-windows beta in the store (hopefully very close now), I'll gladly see if there's any way I can contribute with mwoffliner. I didn't realize it was JavaScript.

**Comment by Jaifroid** [comment]
Interestingly, this also typesets equations in the non-image Wikipedia ZIMs. Screenshot is from the article "List of integrals of trigonometric functions", which has masses and mases of equations, in `wikipedia-all-en-nopic-2017-01` (this is Kiwix JS Windows in dark mode on a Lumia 950XL Windows Mobile):

![image](https://user-images.githubusercontent.com/4304337/29031601-b72a10a0-7b86-11e7-9745-dceee6b1ff72.png)

**Comment by Jaifroid** [comment]
I've solved the FFOS slowness with MathJax by using one of the combined configuration files they provide for the purpose of speeding up loading. This increases loading/rendering speed to near instantaneous on all my testing platforms even on heavy pages full of formulae! Try, for example, Pythagorean Theorem in English Wikipedia, with or without images, or for an extreme case the above List of Integrals.

**Comment by Jaifroid** [comment]
Please see #308 for a MathJax solution to rendering Mathoverflow equations in Kiwix JS Windows.

My thoughts on backporting such a solution to Kiwix JS:

1. In theory, MathJax support in Stackexchange-family ZIMs can be had "for free" if we can get JavaScript working in the app. I tried this method first in Kiwix JS Windows, as it seemed to me to be more "universal", but it didn't turn out to be straightforward. In the end, as I already have a MathJax implementation in Kiwix JS Windows, it seemed more straightforward simply to hook into the implementation.

2. An advantage of the method I chose is that Wikipedia does not provide the MathJax JavaScript necessary to render equations that are given as TeX strings, but prefers the solution of fallback images. With our current decompression technology, most browswers cannot cope with the number of fallback SVGs that appear on a page like Sine (see #278), and most apps simply crash due to memory solution. The fastest solution and the least memory intensive is to provide the MathJax implementation in the app.

3. Maybe we can change the decompression technology (see #219) and avoid the issue that way. However, IMHO it will still be slow, and battery-draining to decompress the hundreds of SVGs that appear on a page like Sine.html. You can see this even in the x86 app with its native C++ decompression.

In sum, any attempt to backport can only be undertaken after taking a decision on the desired direction for Kiwix JS.

**Comment by Jaifroid** [comment]
I think this issue has been superseded by the much faster extraction of SVGs we achieved when re-compiling xzdec, and although Kiwix JS cannot support MathML in Stackexchange ZIMs (where no SVG is provided) in jQuery mode, we provide a banner asking users to switch to SW mode. If they cannot, they can use Kiwix JS Windows, which has KaTeX integrated.
